### DISCUSSION TITLE
Setting up Joined Table Inheritance on imperativelly mapped `attrs.define`d classes produces wrong SQL

### DISCUSSION THREAD

**Discussion by tribals** [sqlalchemy/sqlalchemy#11573] [Usage Questions] [answered]
Hello!

I'm trying to set up "joined table inheritance" on imperatively mapped classes that written with `attrs`. Here I've ported example from docs to equivalent classes in `attrs`:

```python
import attrs


@attrs.define(slots=False)
class Employee:
    pk: int
    name: str
    kind: str


@attrs.define(slots=False)
class Engineer(Employee):
    engineer_name: str


@attrs.define(slots=False)
class Manager(Employee):
    manager_name: str


from sqlalchemy.schema import Table, Column, ForeignKey, MetaData
from sqlalchemy.types import Integer, String

from sqlalchemy.orm import registry


MD = MetaData()

employees = Table(
    'employees', MD,
    Column('pk', Integer, primary_key=True),
    Column('name', String),
    Column('kind', String),
)

engineers = Table(
    'engineers', MD,
    Column('pk', ForeignKey('employees.pk'), primary_key=True),
    Column('engineer_name', String),
)
managers = Table(
    'managers', MD,
    Column('pk', ForeignKey('employees.pk'), primary_key=True),
    Column('manager_name', String),
)

reg = registry(metadata=MD)

reg.map_imperatively(Employee, employees, polymorphic_on=employees.c.kind)
reg.map_imperatively(Engineer, engineers, polymorphic_identity='engineer')
reg.map_imperatively(Manager, managers, polymorphic_identity='manager')


from sqlalchemy.sql import select

query = select(Engineer).where(Employee.name == 'Michael')
```

This results in such query:

```sql
SELECT engineers.pk, engineers.engineer_name
FROM engineers, employees
WHERE employees.name = :name_1
```

I'm trying to do so in order to separate database schema description from "business logic" - to have "entities" in terms of DDD - endowed with that "business logic" thing.

**Comment by CaselIT** [comment]
Hi,

You are not setting polymorphic_identity in the employee.

@zzzeek maybe this should raise an error when a class that's not abstract has no polymorphic_identity?

**Reply by tribals** [reply]
According to docs, it is optional:

> The base class in a joined inheritance hierarchy is configured with additional arguments that will indicate to the polymorphic discriminator column, and optionally a polymorphic identifier for the base class itself:

[Ref.](https://docs.sqlalchemy.org/en/20/orm/inheritance.html#joined-table-inheritance:~:text=polymorphic%20discriminator%20column%2C-,and%20optionally,-a%20polymorphic%20identifier)

**Reply by tribals** [reply]
Also, tried to set it - did not work.

**Reply by tribals** [reply]
Also - I don't need to set it, actually - table for base class contains only *external identity* of entity, and it's "state" - discriminator column. There should be no case when one tries to "create" such a base - it is pointless.

**Reply by tribals** [reply]
But - I *need* to include some attributes of base into children - that's the whole point of inheritance, in OOP.

**Reply by zzzeek** [reply]
> You are not setting polymorphic_identity in the employee.
> 
> @zzzeek maybe this should raise an error when a class that's not abstract has no polymorphic_identity?

that's not the issue here, see my answer below

**Comment by zzzeek** [comment, marked as answer]
these imperative mappings lack the `inherits` parameter so SQLAlchemy knows nothing about how you want these to inherit.  I think you mean to specify like this:

```py
reg.map_imperatively(Employee, employees, polymorphic_on=employees.c.kind)
reg.map_imperatively(Engineer, engineers, polymorphic_identity='engineer', inherits=Employee)
reg.map_imperatively(Manager, managers, polymorphic_identity='manager', inherits=Employee)

```

**Reply by CaselIT** [reply]
would it make sense to raise/warn when inherits is missing?

**Reply by zzzeek** [reply]
map_imperatively() is for the power user that wants to tinker with things, and it is legal to map a subclass without using inherits, that's actually a use case

**Reply by CaselIT** [reply]
oh interesting. never used inheritance much

**Reply by tribals** [reply]
Will try and report later.

**Reply by tribals** [reply]
Works like a charm! I can try to make it on the way to docs, in either "Mapping impratively" section, or "Mapping attrs with imperative mapping".

**Reply by tribals** [reply]
Yaaay! Now that my *Repositories* got actual *responsibility* (to make themselves **sound** in codebase!) And there is no more dichotomy about where one should set query options - on "declarative models", or "at query time, with `.options()`" - it is same place now! ^_^

```python
_reg = registry(metadata=METADATA)


E = TypeVar('E', bound=order.Identity)


class Repository(ABC, Generic[E]):
    @abstractmethod
    def add(self, session: Session, entity: E) -> None: ...


class AcceptedOrders(Repository[order.Accepted]):
    _reg.map_imperatively(
        order.Identity,
        T_orders,
        polymorphic_on=T_orders.c.state,
    )
    _reg.map_imperatively(
        order.Accepted,
        T_accepted_orders,
        inherits=order.Identity,
        polymorphic_identity=OrderStates.ACCEPTED,
        properties=dict(
            product_items=relationship(order.ProductItem, lazy='joined'),
        ),
    )
    _reg.map_imperatively(order.ProductItem, T_product_items)

    def add(self, session: Session, order: E) -> None:
        # session.execute()
        session.commit()

    def get_by_order_id(self, session: Session, order_id: OrderId) -> order.Accepted:
        query = select(order.Accepted).where(order.Identity.order_id == order_id)

        result: ScalarResult = session.scalars(query)

        # breakpoint()
        return result.unique().one()
```

Really bright code!

**Reply by CaselIT** [reply]
What you are doing is equivalent to "on declarative model", but if you prefer doing it like this then by all means

**Reply by tribals** [reply]
Despite many other advantages like conducting SPR in *Repositories* this approach has one major: there is no "creepy dunders" on models classes, and *everything set with only Python syntax,* without resorting to "strings" and the like!

**Comment by tribals** [comment]
Also, I've faced sporadic issues of "entity being already mapped", which I believe caused by the way pytest doing imports - is it safe to  call `map_imperatively` multiple times?

**Reply by CaselIT** [reply]
in general no on the same classes. You can map the table multiple times instead
